app/models/location_report.rb

```diff
@@ -0,0 +1,208 @@
+# This file is part of SEQUENCESCAPE; it is distributed under the terms of
+# GNU General Public License version 1 or later;
+# Please refer to the LICENSE and README files for information on licensing and
+# authorship of this file.
+# Copyright (C) 2018 Genome Research Ltd.
+
+class LocationReport < ApplicationRecord
+  extend DbFile::Uploader
+
+  attr_accessor :barcodes_list
+
+  def barcodes_text
+    barcodes_list.join(' ') unless @barcodes_list.nil?
+  end
+
+  def barcodes_text=(value)
+    self.barcodes_list = if value.nil?
+                           nil
+                         else
+                           value.strip.split(/[\s]+/)
+                         end
+  end
+
+  self.per_page = 10
+
+  scope :for_user, ->(user) { where(user_id: user.id) }
+
+  belongs_to :study
+  belongs_to :user
+
+  has_uploaded :report, serialization_column: 'report_filename'
+
+  def headers
+    %w[Barcode HumanBarcode Type Created Location Service Study Owner]
+  end
+
+  def generate!
+    ActiveRecord::Base.transaction do
+      csv_options = { row_sep: "\r\n", force_quotes: true }
+      filename    = "location_report_#{name}.csv"
+      Tempfile.open(filename) do |tempfile|
```

Comment:
http://ruby-doc.org/stdlib-2.4.1/libdoc/csv/rdoc/CSV.html
Using a tempfile is probably going to be necessary still, but you might be able to wrap it in the CSV library.
Not sure whet the re-opening is about. Suggest getting rid of it and making sure everything still passes/works as expected.